jobsdb/internal/dsindex/dsindex.go

```diff
@@ -0,0 +1,128 @@
+package dsindex
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// Index represents a sortable dataset index, e.g. 1 < 1_1 < 1_1_1 < 1_2 < 2
+type Index struct {
+	segments []int
+}
+
+// MustBump returns the next index that is greater than the current one,
+// but still less than the index provided in the parameter. Panics on error
+func (idx *Index) MustBump(previous *Index) *Index {
+	res, err := idx.Bump(previous)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// Bump returns the next index that is greater than the current one,
+// but still less than the index provided in the parameter.
+//
+// Bump doesn't increment the first, major segment of the index, it starts from the second segment instead.
+func (idx *Index) Bump(before *Index) (*Index, error) {
+	if !idx.Less(before) {
+		return nil, fmt.Errorf("%s is not before %s", idx, before)
+	}
+	var segment int // never increasing the major segment (index 0)
+	for {
+		segment++
+		res, err := idx.Increment(segment)
+		if err != nil {
+			return nil, err
+		}
+		pl := idx.Less(res)
+		rl := res.Less(before)
+		if pl && rl {
+			return res, nil
+		}
+	}
+}
+
+// MustIncrement returns a new dataset index that is incremented by one in the specified segment, panics on error
+func (idx *Index) MustIncrement(segment int) *Index {
+	res, err := idx.Increment(segment)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// Increment returns a new dataset index that is incremented by one in the specified segment
+func (idx *Index) Increment(segment int) (*Index, error) {
+	if segment < 0 || segment > idx.Length() {
+		return nil, fmt.Errorf("cannot increment segment %d of %s", segment, idx)
+	}
+	var res *Index = &Index{}
+	for i := 0; i < segment; i++ {
+		toAppend := idx.segments[i]
+		res.segments = append(res.segments, toAppend)
+	}
+	var lastSegment int
+	if idx.Length() > segment {
+		lastSegment = idx.segments[segment]
+	}
+	res.segments = append(res.segments, lastSegment+1)
+	return res, nil
+}
+
+// Less returns true if this dataset index is Less than the other dataset index
+func (idx *Index) Less(other *Index) bool {
```

Comment:
Since comparing or bumping a nil Index cannot yield any meaningful result other than a panic, no extra branch is needed for checking against it. A nil pointer dereference panic is guaranteed and should be enough :)